measureme/src/stringtable.rs

```diff
@@ -233,41 +267,63 @@ impl<S: SerializationSink> StringTableBuilder<S> {
         StringTableBuilder {
             data_sink,
             index_sink,
-            id_counter: AtomicU32::new(METADATA_STRING_ID + 1),
         }
     }
 
-    pub fn alloc_with_reserved_id<STR: SerializableString + ?Sized>(
-        &self,
-        id: StringId,
-        s: &STR,
-    ) -> StringId {
-        assert!(id.0 <= MAX_PRE_RESERVED_STRING_ID);
-        self.alloc_unchecked(id, s);
-        id
+    /// Creates a mapping so that `virtual_id` will resolve to the contents of
+    /// `concrete_id` when reading the string table.
+    pub fn map_virtual_to_concrete_string(&self, virtual_id: StringId, concrete_id: StringId) {
+        // This assertion does not use `is_virtual` on purpose because that
+        // would also allow to overwrite `METADATA_STRING_ID`.
+        assert!(virtual_id.0 <= MAX_USER_VIRTUAL_STRING_ID);
+        serialize_index_entry(&*self.index_sink, virtual_id, concrete_id.to_addr());
     }
 
-    pub(crate) fn alloc_metadata<STR: SerializableString + ?Sized>(&self, s: &STR) -> StringId {
-        let id = StringId(METADATA_STRING_ID);
-        self.alloc_unchecked(id, s);
-        id
+    pub fn bulk_map_virtual_to_single_concrete_string<I>(
+        &self,
+        virtual_ids: I,
+        concrete_id: StringId,
+    ) where
+        I: Iterator<Item = StringId> + ExactSizeIterator,
+    {
+        // TODO: Index data encoding could have special bulk mode that assigns
+        //       multiple StringIds to the same addr, so we don't have to repeat
+        //       the `concrete_id` over and over.
+
+        type MappingEntry = (u32, u32);
+        assert!(std::mem::size_of::<MappingEntry>() == 8);
+
+        let to_addr_le = concrete_id.to_addr().0.to_le();
+
+        let serialized: Vec<MappingEntry> = virtual_ids
+            .map(|from| {
+                let id = from.0;
+                assert!(id <= MAX_USER_VIRTUAL_STRING_ID);
+                (id.to_le(), to_addr_le)
+            })
+            .collect();
+
+        let num_bytes = serialized.len() * std::mem::size_of::<MappingEntry>();
+        let byte_ptr = serialized.as_ptr() as *const u8;
+
+        let bytes = unsafe { std::slice::from_raw_parts(byte_ptr, num_bytes) };
```

Comment:
According to [this](https://doc.rust-lang.org/reference/type-layout.html#tuple-layout) tuples don't have a defined layout so I don't think this is technically allowed? Switching to a `[u32; 2]` array should be ok.

Reply:
Oh! Great catch!

Reply:
Updated.